### PR TITLE
ENH: demux-single

### PR DESCRIPTION
**Pull request by thermokarst**
~Depends on https://github.com/qiime2/q2-types/pull/163~

Part of the ~umbrella~ web of work related to https://github.com/qiime2/q2-dada2/issues/32

---

TODO:

- [x] Expose error tolerance
- [x] Add additional tests as requested below
- [x] Add citation
- [x] Update plugin description as per @nbokulich's suggestion
- [x] Add some additional edge-case tests (all reads match; no reads match). These shouldn't be a big deal, since the `FastqGzFormat` supports empty files already, so no special handling is required.
- [x] Remove visualization? I discussed this viz with @ebolyen and he pointed out that it is probably of limited utility (and can potentially be huge - it will be at least as big as the input sequences themselves)

### CODE REVIEW

File: q2_cutadapt/plugin_setup.py [hunk 1]
```diff
@@ -6,18 +6,67 @@
...
     name='cutadapt',
     version=q2_cutadapt.__version__,
     website='https://github.com/qiime2/q2-cutadapt',
     package='q2_cutadapt',
-    description='This QIIME 2 plugin supports removing adapters, primers, '
-                'and other unwanted sequences from sequence data.',
-    short_description='Plugin for removing unwanted sequences from sequence '
+    description='This QIIME 2 plugin uses cutadapt to work with '
+                'non-biological signal in sequences (e.g. barcodes, primers).',
+    short_description='Plugin for handling non-biological signal in sequence '
```
Comment (nbokulich) [inline q2_cutadapt/plugin_setup.py]: Description is a bit unclear/overly specific: "Plugin for handling non-biological signal in sequence data". What about trimming primers, etc — those would still be biological. Maybe (to condense a sentence from the cutadapt docs), something like "plugin for removing adapter sequences, primers, and other unwanted sequence from sequence data"

**Comment by thermokarst** [comment]
Thanks @nbokulich --- we definitely have a bit of a dichotomy when it comes to plugins --- we either have topic-specific plugins (e.g. `q2-sample-classifier`, `q2-diversity`) vs. tool-specific plugins (e.g. `q2-dada2`, `q2-vsearch`). I am fine with it in either plugin (`q2-cutadapt` vs `q2-demux`), but to me it felt like it made more sense in the `q2-cutadapt` plugin.

@gregcaporaso, @ebolyen, @jairideout, care to comment on this?

**Comment by gregcaporaso** [comment]
I also think it makes sense to keep this in `cutadapt` with a possible alias at some point in the future.

File: q2_cutadapt/tests/test_demux.py [hunk 2]
```diff
@@ -6,17 +6,185 @@
...
+            self.assertTrue(sample_id in filename)
+            self.assertTrue(barcode in filename)
+
+        obs_untrimmed = obs_untrimmed_art.view(
+            MultiplexedSingleEndBarcodeInSequenceDirFmt)
+        obs_untrimmed = obs_untrimmed.file.view(FastqGzFormat)
+        obs_untrimmed = gzip.decompress(obs_untrimmed.path.read_bytes())
+        self.assertEqual(b'@id6\nGGGGACGTACGT\n+\nzzzzzzzzzzzz\n',
+                         obs_untrimmed)
+
```
Comment (ebolyen) [inline q2_cutadapt/tests/test_demux.py]: Add a test for having an extra barcode in the metadata column

Comment (ebolyen) [reply q2_cutadapt/tests/test_demux.py]: Also maybe a test for variable length barcodes?

Comment (thermokarst) [reply q2_cutadapt/tests/test_demux.py]: Sounds good! Also need to expose the error tolerance, will add some tests for that, too.
